File: agent/cursor/src/store.ts
import { randomUUID } from "node:crypto";

import type { JsonObject } from "@bufbuild/protobuf";

import {
  AgentExecutionStatus,
  AgentSessionState,
  AgentToolSourceMode,
  type AgentMessage,
  type ListedAgentTool,
  type AgentToolRef,
  type AgentSession,
  type AgentTurnEvent,
  type AgentTurn,
  type AgentTurnOutput,
} from "@valon-technologies/gestalt/services/agent";

export type PreparedWorkspace = {
  root: string;
  cwd: string;
};

export const TERMINAL_STATUSES = new Set<AgentExecutionStatus>([
  AgentExecutionStatus.SUCCEEDED,
  AgentExecutionStatus.FAILED,
  AgentExecutionStatus.CANCELED,
]);

export const SESSION_VISIBILITY_PRIVATE = "private";
export const SESSION_VISIBILITY_COMPANY = "company";

export type StoredSessionVisibility =
  | typeof SESSION_VISIBILITY_PRIVATE
  | typeof SESSION_VISIBILITY_COMPANY;

export class StoreConflictError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StoreConflictError";
  }
}

export type StoredSession = {
  sessionId: string;
  idempotencyKey: string;
  providerName: string;
  model: string;
  clientRef: string;
  state: AgentSessionState;
  metadata: Record<string, unknown>;
  visibility: StoredSessionVisibility;
  preparedWorkspace: PreparedWorkspace | undefined;
  toolSource: AgentToolSourceMode;
  toolRefs: AgentToolRef[];
  listedTools: ListedAgentTool[];
  createdBySubjectId: string;
  createdAt: Date;
  updatedAt: Date;
  lastTurnAt: Date | undefined;
};

export type StoredTurn = {
  turnId: string;
  sessionId: string;
  idempotencyKey: string;
  providerName: string;
  model: string;
  status: AgentExecutionStatus;
  messages: AgentMessage[];
  output: AgentTurnOutput | undefined;
  statusMessage: string;
  createdBySubjectId: string;
  createdAt: Date;
  startedAt: Date | undefined;
  completedAt: Date | undefined;
  executionRef: string;
};

export type StoredTurnEvent = {
  eventId: string;
  turnId: string;
  seq: bigint;
  eventType: string;
  source: string;
  visibility: string;
  data: Record<string, unknown>;
  createdAt: Date;
};

export class InMemoryRunStore {
  private readonly sessions = new Map<string, StoredSession>();
  private readonly sessionIdempotency = new Map<string, string>();
  private readonly turns = new Map<string, StoredTurn>();
  private readonly turnIdempotency = new Map<string, string>();
  private readonly events = new Map<string, StoredTurnEvent[]>();

  close(): void {
    this.sessions.clear();
    this.sessionIdempotency.clear();
    this.turns.clear();
    this.turnIdempotency.clear();
    this.events.clear();
  }

  createSession(input: {
    idempotencyKey: string;
    providerName: string;
    model: string;
    clientRef: string;
    metadata: Record<string, unknown>;
    visibility: StoredSessionVisibility;
    preparedWorkspace: PreparedWorkspace | undefined;
    toolSource: AgentToolSourceMode;
    toolRefs: readonly AgentToolRef[];
    listedTools: readonly ListedAgentTool[];
    createdBySubjectId: string;
  }): { session: StoredSession; created: boolean } {
    const dedupKey = sessionDedupKey(input.createdBySubjectId, input.idempotencyKey);
    if (input.idempotencyKey.trim()) {
      const existingId = this.sessionIdempotency.get(dedupKey);
      if (existingId) {
        const idempotent = this.sessions.get(existingId);
        if (idempotent) {
          return { session: cloneSession(idempotent), created: false };
        }
      }
    }
    const sessionId = randomUUID();
    const now = new Date();
    const session: StoredSession = {
      sessionId,
      idempotencyKey: input.idempotencyKey,
      providerName: input.providerName,
      model: input.model,
      clientRef: input.clientRef,
      state: AgentSessionState.ACTIVE,
      metadata: cloneRecord(input.metadata),
      visibility: input.visibility,
      preparedWorkspace: cloneMaybe(input.preparedWorkspace),
      toolSource: input.toolSource,
      toolRefs: cloneToolRefs(input.toolRefs),
      listedTools: cloneListedTools(input.listedTools),
      createdBySubjectId: input.createdBySubjectId.trim(),
      createdAt: now,
      updatedAt: now,
      lastTurnAt: undefined,
    };
    this.sessions.set(sessionId, session);
    if (input.idempotencyKey.trim()) {
      this.sessionIdempotency.set(dedupKey, sessionId);
    }
    return { session: cloneSession(session), created: true };
  }

  getSession(sessionId: string): StoredSession | undefined {
    const session = this.sessions.get(sessionId.trim());
    return session ? cloneSession(session) : undefined;
  }

  getSessionByIdempotencyKey(
    createdBySubjectId: string,
    idempotencyKey: string,
  ): StoredSession | undefined {
    const sessionId = this.sessionIdempotency.get(
      sessionDedupKey(createdBySubjectId, idempotencyKey),
    );
    if (!sessionId) {
      return undefined;
    }
    const session = this.sessions.get(sessionId);
    return session ? cloneSession(session) : undefined;
  }

  listSessions(input: {
    sessionIds?: readonly string[];
    subjectId: string | undefined;
    state: AgentSessionState | undefined;
    limit: number | undefined;
  }): StoredSession[] {
    const requested = (input.sessionIds ?? []).map((value) => value.trim()).filter(Boolean);
    let sessions = requested.length
      ? requested.flatMap((id) => {
          const session = this.sessions.get(id);
          return session ? [session] : [];
        })
      : [...this.sessions.values()];
    if (input.subjectId) {
      sessions = sessions.filter(
        (session) =>
          session.createdBySubjectId === input.subjectId ||
          session.visibility === SESSION_VISIBILITY_COMPANY,
      );
    } else {
      sessions = [];
    }
    if (input.state) {
      sessions = sessions.filter((session) => session.state === input.state);
    }
    sessions = sessions.sort((left, right) => {
      const leftTime = left.lastTurnAt ?? left.updatedAt;
      const rightTime = right.lastTurnAt ?? right.updatedAt;
      return rightTime.getTime() - leftTime.getTime();
    });
    if (input.limit && input.limit > 0) {
      sessions = sessions.slice(0, input.limit);
    }
    return sessions.map(cloneSession);
  }

  updateSession(input: {
    sessionId: string;
    clientRef: string | undefined;
    state: AgentSessionState | undefined;
    metadata: Record<string, unknown> | undefined;
  }): StoredSession | undefined {
    const session = this.sessions.get(input.sessionId.trim());
    if (!session) {
      return undefined;
    }
    if (input.clientRef) {
      session.clientRef = input.clientRef;
    }
    if (input.state) {
      session.state = input.state;
    }
    if (input.metadata !== undefined) {
      session.metadata = cloneRecord(input.metadata);
    }
    session.updatedAt = new Date();
    return cloneSession(session);
  }

  beginTurn(input: {
    turnId: string;
    sessionId: string;
    idempotencyKey: string;
    providerName: string;
    model: string;
    messages: readonly AgentMessage[];
    createdBySubjectId: string;
    executionRef: string;
  }): { turn: StoredTurn; created: boolean } {
    const turnId = input.turnId.trim();
    const sessionId = input.sessionId.trim();
    if (!turnId) {
      throw new Error("turn_id is required");
    }
    if (!sessionId) {
      throw new Error("session_id is required");
    }
    const existing = this.turns.get(turnId);
    if (existing) {
      if (existing.sessionId !== sessionId) {
        throw new StoreConflictError(`turn_id ${JSON.stringify(turnId)} already exists for another session`);
      }
      return { turn: cloneTurn(existing), created: false };
    }
    const idempotencyKey = `${sessionId}\n${input.idempotencyKey}`;
    if (input.idempotencyKey && this.turnIdempotency.has(idempotencyKey)) {
      const existingId = this.turnIdempotency.get(idempotencyKey);
      const idempotent = existingId ? this.turns.get(existingId) : undefined;
      if (idempotent) {
        return { turn: cloneTurn(idempotent), created: false };
      }
    }
    const now = new Date();
    const turn: StoredTurn = {
      turnId,
      sessionId,
      idempotencyKey: input.idempotencyKey,
      providerName: input.providerName,
      model: input.model,
      status: AgentExecutionStatus.RUNNING,
      messages: cloneMessages(input.messages),
      output: undefined,
      statusMessage: "",
      createdBySubjectId: input.createdBySubjectId.trim(),
      createdAt: now,
      startedAt: now,
      completedAt: undefined,
      executionRef: input.executionRef,
    };
    this.turns.set(turnId, turn);
    if (input.idempotencyKey) {
      this.turnIdempotency.set(idempotencyKey, turnId);
    }
    const session = this.sessions.get(sessionId);
    if (session) {
      session.lastTurnAt = now;
      session.updatedAt = now;
    }
    this.appendEvent({
      turnId,
      eventType: "turn.started",
      source: input.providerName,
      data: { model: input.model },
    });
    return { turn: cloneTurn(turn), created: true };
  }

  getTurn(turnId: string): StoredTurn | undefined {
    const turn = this.turns.get(turnId.trim());
    return turn ? cloneTurn(turn) : undefined;
  }

  listTurns(input: {
    sessionId: string;
    turnIds: readonly string[] | undefined;
    subjectId: string | undefined;
    status: AgentExecutionStatus | undefined;
    limit: number | undefined;
  }): StoredTurn[] {
    const requested = (input.turnIds ?? []).map((value) => value.trim()).filter(Boolean);
    const sessionId = input.sessionId.trim();
    let turns = requested.length
      ? requested.flatMap((id) => {
          const turn = this.turns.get(id);
          return turn ? [turn] : [];
        })
      : [...this.turns.values()];
    if (sessionId) {
      turns = turns.filter((turn) => turn.sessionId === sessionId);
    } else if (!requested.length) {
      turns = [];
    }
    if (input.subjectId) {
      turns = turns.filter((turn) => turn.createdBySubjectId === input.subjectId);
    }
    if (input.status) {
      turns = turns.filter((turn) => turn.status === input.status);
    }
    turns = turns.sort((left, right) => right.createdAt.getTime() - left.createdAt.getTime());
    if (input.limit && input.limit > 0) {
      turns = turns.slice(0, input.limit);
    }
    return turns.map(cloneTurn);
  }

  completeTurn(turnId: string, output: AgentTurnOutput): StoredTurn | undefined {
    const turn = this.turns.get(turnId.trim());
    if (!turn || TERMINAL_STATUSES.has(turn.status)) {
      return turn ? cloneTurn(turn) : undefined;
    }
    turn.status = AgentExecutionStatus.SUCCEEDED;
    turn.output = cloneMaybe(output);
    turn.completedAt = new Date();
    this.appendEvent({
      turnId: turn.turnId,
      eventType: "assistant.message",
      source: turn.providerName,
      data: assistantMessageData(output),
    });
    this.appendEvent({
      turnId: turn.turnId,
      eventType: "turn.completed",
      source: turn.providerName,
      data: { status: "succeeded" },
    });
    return cloneTurn(turn);
  }

  failTurn(turnId: string, message: string): StoredTurn | undefined {
    const turn = this.turns.get(turnId.trim());
    if (!turn || TERMINAL_STATUSES.has(turn.status)) {
      return turn ? cloneTurn(turn) : undefined;
    }
    turn.status = AgentExecutionStatus.FAILED;
    turn.statusMessage = message;
    turn.completedAt = new Date();
    this.appendEvent({
      turnId: turn.turnId,
      eventType: "turn.failed",
      source: turn.providerName,
      data: { error: message },
    });
    return cloneTurn(turn);
  }

  cancelTurn(turnId: string, reason: string): StoredTurn | undefined {
    const turn = this.turns.get(turnId.trim());
    if (!turn) {
      return undefined;
    }
    if (!TERMINAL_STATUSES.has(turn.status)) {
      turn.status = AgentExecutionStatus.CANCELED;
      turn.statusMessage = reason;
      turn.completedAt = new Date();
      this.appendEvent({
        turnId: turn.turnId,
        eventType: "turn.canceled",
        source: turn.providerName,
        data: { reason },
      });
    }
    return cloneTurn(turn);
  }

  appendEvent(input: {
    turnId: string;
    eventType: string;
    source: string;
    data: Record<string, unknown>;
  }): StoredTurnEvent {
    const events = this.events.get(input.turnId) ?? [];
    this.events.set(input.turnId, events);
    const event: StoredTurnEvent = {
      eventId: `${input.turnId}:${events.length + 1}`,
      turnId: input.turnId,
      seq: BigInt(events.length + 1),
      eventType: input.eventType,
      source: input.source,
      visibility: "external",
      data: cloneRecord(input.data),
      createdAt: new Date(),
    };
    events.push(event);
    return cloneEvent(event);
  }

  listTurnEvents(input: { turnId: string; afterSeq?: bigint; limit?: number }): StoredTurnEvent[] {
    const afterSeq = input.afterSeq ?? 0n;
    let events = (this.events.get(input.turnId.trim()) ?? []).filter(
      (event) => event.seq > afterSeq,
    );
    if (input.limit && input.limit > 0) {
      events = events.slice(0, input.limit);
    }
    return events.map(cloneEvent);
  }
}

export function sessionToAgentSession(session: StoredSession, summaryOnly = false): AgentSession {
  const out: AgentSession = {
    id: session.sessionId,
    providerName: session.providerName,
    model: session.model,
    clientRef: session.clientRef,
    state: session.state,
    createdBySubjectId: session.createdBySubjectId,
    createdAt: new Date(session.createdAt),
    updatedAt: new Date(session.updatedAt),
  };
  if (!summaryOnly && Object.keys(session.metadata).length > 0) {
    out.metadata = session.metadata as JsonObject;
  }
  if (session.lastTurnAt) {
    out.lastTurnAt = new Date(session.lastTurnAt);
  }
  return out;
}

export function turnToAgentTurn(turn: StoredTurn, summaryOnly = false): AgentTurn {
  const out: AgentTurn = {
    id: turn.turnId,
    sessionId: turn.sessionId,
    providerName: turn.providerName,
    model: turn.model,
    status: turn.status,
    messages: summaryOnly ? [] : cloneMessages(turn.messages),
    output: summaryOnly
      ? { case: undefined }
      : (cloneMaybe(turn.output) ?? { case: undefined }),
    statusMessage: turn.statusMessage,
    createdBySubjectId: turn.createdBySubjectId,
    executionRef: turn.executionRef,
    createdAt: new Date(turn.createdAt),
  };
  if (turn.startedAt) {
    out.startedAt = new Date(turn.startedAt);
  }
  if (turn.completedAt) {
    out.completedAt = new Date(turn.completedAt);
  }
  return out;
}

export function turnEventToAgentTurnEvent(event: StoredTurnEvent): AgentTurnEvent {
  return {
    id: event.eventId,
    turnId: event.turnId,
    seq: event.seq,
    type: event.eventType,
    source: event.source,
    visibility: event.visibility,
    data: event.data as JsonObject,
    createdAt: new Date(event.createdAt),
  };
}

function sessionDedupKey(createdBySubjectId: string, idempotencyKey: string): string {
  return `${createdBySubjectId.trim()}${idempotencyKey.trim()}`;
}

export function cloneRecord(value: Record<string, unknown>): Record<string, unknown> {
  return structuredClone(value);
}

function cloneSession(session: StoredSession): StoredSession {
  const copy: StoredSession = {
    ...session,
    metadata: cloneRecord(session.metadata),
    preparedWorkspace: cloneMaybe(session.preparedWorkspace),
    toolRefs: cloneToolRefs(session.toolRefs),
    listedTools: cloneListedTools(session.listedTools),
    createdBySubjectId: session.createdBySubjectId,
    createdAt: new Date(session.createdAt),
    updatedAt: new Date(session.updatedAt),
  };
  if (session.lastTurnAt) {
    copy.lastTurnAt = new Date(session.lastTurnAt);
  }
  return copy;
}

function cloneTurn(turn: StoredTurn): StoredTurn {
  const copy: StoredTurn = {
    ...turn,
    messages: cloneMessages(turn.messages),
    output: cloneMaybe(turn.output),
    createdBySubjectId: turn.createdBySubjectId,
    createdAt: new Date(turn.createdAt),
  };
  if (turn.startedAt) {
    copy.startedAt = new Date(turn.startedAt);
  }
  if (turn.completedAt) {
    copy.completedAt = new Date(turn.completedAt);
  }
  return copy;
}

function cloneEvent(event: StoredTurnEvent): StoredTurnEvent {
  return {
    ...event,
    data: cloneRecord(event.data),
    createdAt: new Date(event.createdAt),
  };
}

function cloneMessages(messages: readonly AgentMessage[]): AgentMessage[] {
  return structuredClone([...messages]);
}

function cloneToolRefs(refs: readonly AgentToolRef[]): AgentToolRef[] {
  return structuredClone([...refs]);
}

function cloneListedTools(tools: readonly ListedAgentTool[]): ListedAgentTool[] {
  return structuredClone([...tools]);
}

function cloneMaybe<T>(value: T | undefined): T | undefined {
  return value === undefined ? undefined : structuredClone(value);
}

function assistantMessageData(output: AgentTurnOutput): Record<string, unknown> {
  if (output.case === "text") {
    return { text: output.value.text };
  }
  if (output.case === "structured") {
    const data: Record<string, unknown> = { text: output.value.text };
    if (output.value.value !== undefined) {
      data.value = cloneMaybe(output.value.value);
    }
    return data;
  }
  throw new Error("completed turn output must include text or structured");
}

File: agent/codex/internals/store.py
from __future__ import annotations

import copy
import threading
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

import gestalt

TERMINAL_STATUSES = {
    gestalt.AGENT_EXECUTION_STATUS_SUCCEEDED,
    gestalt.AGENT_EXECUTION_STATUS_FAILED,
    gestalt.AGENT_EXECUTION_STATUS_CANCELED,
}

SESSION_VISIBILITY_COMPANY = "company"
SESSION_VISIBILITY_PRIVATE = "private"


class StoreConflictError(ValueError):
    pass


@dataclass(slots=True)
class StoredSession:
    session_id: str
    idempotency_key: str
    provider_name: str
    model: str
    client_ref: str
    state: int
    metadata: dict[str, Any]
    prepared_workspace: dict[str, str] | None
    created_by: dict[str, str]
    visibility: str
    created_at: datetime
    updated_at: datetime
    last_turn_at: datetime | None = None


@dataclass(slots=True)
class StoredTurn:
    turn_id: str
    session_id: str
    idempotency_key: str
    provider_name: str
    model: str
    status: int
    messages: list[dict[str, Any]]
    output: gestalt.AgentTurnOutput | None
    status_message: str
    created_by: dict[str, str]
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    execution_ref: str


@dataclass(slots=True)
class StoredTurnEvent:
    event_id: str
    turn_id: str
    seq: int
    event_type: str
    source: str
    visibility: str
    data: dict[str, Any]
    created_at: datetime


class InMemoryRunStore:
    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._sessions: dict[str, StoredSession] = {}
        self._session_idempotency: dict[str, str] = {}
        self._turns: dict[str, StoredTurn] = {}
        self._turn_idempotency: dict[tuple[str, str], str] = {}
        self._events: dict[str, list[StoredTurnEvent]] = {}

    def close(self) -> None:
        with self._lock:
            self._sessions.clear()
            self._session_idempotency.clear()
            self._turns.clear()
            self._turn_idempotency.clear()
            self._events.clear()

    def create_session(
        self,
        *,
        session_id: str,
        idempotency_key: str,
        provider_name: str,
        model: str,
        client_ref: str,
        metadata: dict[str, Any],
        prepared_workspace: dict[str, str] | None,
        created_by: dict[str, str],
    ) -> tuple[StoredSession, bool]:
        session_id = session_id.strip()
        if not session_id:
            raise ValueError("session_id is required")
        with self._lock:
            if session_id in self._sessions:
                return copy.deepcopy(self._sessions[session_id]), False
            if idempotency_key and idempotency_key in self._session_idempotency:
                return copy.deepcopy(self._sessions[self._session_idempotency[idempotency_key]]), False
            now = _utcnow()
            session = StoredSession(
                session_id=session_id,
                idempotency_key=idempotency_key,
                provider_name=provider_name,
                model=model,
                client_ref=client_ref,
                state=gestalt.AGENT_SESSION_STATE_ACTIVE,
                metadata=copy.deepcopy(metadata),
                prepared_workspace=copy.deepcopy(prepared_workspace) if prepared_workspace is not None else None,
                created_by=dict(created_by),
                visibility=session_visibility_for_create(metadata, created_by),
                created_at=now,
                updated_at=now,
            )
            self._sessions[session_id] = session
            if idempotency_key:
                self._session_idempotency[idempotency_key] = session_id
            return copy.deepcopy(session), True

    def get_session(self, session_id: str) -> StoredSession | None:
        with self._lock:
            session = self._sessions.get(session_id.strip())
            return copy.deepcopy(session) if session is not None else None

    def get_session_by_idempotency_key(self, idempotency_key: str) -> StoredSession | None:
        idempotency_key = idempotency_key.strip()
        if not idempotency_key:
            return None
        with self._lock:
            session_id = self._session_idempotency.get(idempotency_key)
            if not session_id:
                return None
            session = self._sessions.get(session_id)
            return copy.deepcopy(session) if session is not None else None

    def list_sessions(
        self, *, session_ids: list[str] | None = None, subject_id: str = "", state: int = 0, limit: int = 0
    ) -> list[StoredSession]:
        requested_ids = [value.strip() for value in session_ids or [] if value.strip()]
        with self._lock:
            if requested_ids:
                sessions = [self._sessions[value] for value in requested_ids if value in self._sessions]
            else:
                sessions = list(self._sessions.values())
            if not subject_id:
                sessions = []
            else:
                sessions = [session for session in sessions if session_readable_by(session, subject_id)]
            if state:
                sessions = [session for session in sessions if session.state == state]
            sessions = sorted(sessions, key=lambda session: session.last_turn_at or session.updated_at, reverse=True)
            if limit > 0:
                sessions = sessions[:limit]
            return copy.deepcopy(sessions)

    def update_session(
        self, *, session_id: str, client_ref: str = "", state: int = 0, metadata: dict[str, Any] | None = None
    ) -> StoredSession | None:
        with self._lock:
            session = self._sessions.get(session_id.strip())
            if session is None:
                return None
            if client_ref:
                session.client_ref = client_ref
            if state:
                session.state = state
            if metadata is not None:
                session.metadata = copy.deepcopy(metadata)
            session.updated_at = _utcnow()
            return copy.deepcopy(session)

    def begin_turn(
        self,
        *,
        turn_id: str,
        session_id: str,
        idempotency_key: str,
        provider_name: str,
        model: str,
        messages: list[dict[str, Any]],
        created_by: dict[str, str],
        execution_ref: str,
    ) -> tuple[StoredTurn, bool]:
        turn_id = turn_id.strip()
        session_id = session_id.strip()
        if not turn_id:
            raise ValueError("turn_id is required")
        if not session_id:
            raise ValueError("session_id is required")
        with self._lock:
            if turn_id in self._turns:
                existing = self._turns[turn_id]
                if existing.session_id != session_id:
                    raise StoreConflictError(f"turn_id {turn_id!r} already exists for another session")
                return copy.deepcopy(existing), False
            idem_key = (session_id, idempotency_key)
            if idempotency_key and idem_key in self._turn_idempotency:
                existing = self._turns[self._turn_idempotency[idem_key]]
                if existing.session_id != session_id:
                    raise StoreConflictError(f"idempotency_key {idempotency_key!r} already exists for another session")
                return copy.deepcopy(existing), False
            now = _utcnow()
            turn = StoredTurn(
                turn_id=turn_id,
                session_id=session_id,
                idempotency_key=idempotency_key,
                provider_name=provider_name,
                model=model,
                status=gestalt.AGENT_EXECUTION_STATUS_RUNNING,
                messages=copy.deepcopy(messages),
                output=None,
                status_message="",
                created_by=dict(created_by),
                created_at=now,
                started_at=now,
                completed_at=None,
                execution_ref=execution_ref,
            )
            self._turns[turn_id] = turn
            if idempotency_key:
                self._turn_idempotency[idem_key] = turn_id
            if session_id in self._sessions:
                self._sessions[session_id].last_turn_at = now
                self._sessions[session_id].updated_at = now
            self.append_event(turn_id=turn_id, event_type="turn.started", source=provider_name, data={"model": model})
            return copy.deepcopy(turn), True

    def get_turn(self, turn_id: str) -> StoredTurn | None:
        with self._lock:
            turn = self._turns.get(turn_id.strip())
            return copy.deepcopy(turn) if turn is not None else None

    def list_turns(
        self,
        *,
        session_id: str,
        turn_ids: list[str] | None = None,
        subject_id: str = "",
        status: int = 0,
        limit: int = 0,
    ) -> list[StoredTurn]:
        requested_ids = [value.strip() for value in turn_ids or [] if value.strip()]
        with self._lock:
            if requested_ids:
                turns = [self._turns[value] for value in requested_ids if value in self._turns]
            else:
                turns = [turn for turn in self._turns.values() if turn.session_id == session_id.strip()]
            if subject_id:
                turns = [turn for turn in turns if str(turn.created_by.get("subject_id", "")).strip() == subject_id]
            if status:
                turns = [turn for turn in turns if turn.status == status]
            turns = sorted(turns, key=lambda turn: turn.created_at, reverse=True)
            if limit > 0:
                turns = turns[:limit]
            return copy.deepcopy(turns)

    def complete_turn(self, *, turn_id: str, output: gestalt.AgentTurnOutput) -> StoredTurn | None:
        with self._lock:
            turn = self._turns.get(turn_id.strip())
            if turn is None or turn.status in TERMINAL_STATUSES:
                return copy.deepcopy(turn) if turn is not None else None
            turn.status = gestalt.AGENT_EXECUTION_STATUS_SUCCEEDED
            turn.output = copy.deepcopy(output)
            turn.completed_at = _utcnow()
            self.append_event(
                turn_id=turn.turn_id,
                event_type="assistant.message",
                source=turn.provider_name,
                data=_assistant_message_event_data(output),
            )
            self.append_event(
                turn_id=turn.turn_id,
                event_type="turn.completed",
                source=turn.provider_name,
                data={"status": "succeeded"},
            )
            return copy.deepcopy(turn)

    def fail_turn(self, *, turn_id: str, message: str) -> StoredTurn | None:
        with self._lock:
            turn = self._turns.get(turn_id.strip())
            if turn is None or turn.status in TERMINAL_STATUSES:
                return copy.deepcopy(turn) if turn is not None else None
            turn.status = gestalt.AGENT_EXECUTION_STATUS_FAILED
            turn.status_message = message
            turn.completed_at = _utcnow()
            self.append_event(
                turn_id=turn.turn_id, event_type="turn.failed", source=turn.provider_name, data={"error": message}
            )
            return copy.deepcopy(turn)

    def cancel_turn(self, *, turn_id: str, reason: str) -> StoredTurn | None:
        with self._lock:
            turn = self._turns.get(turn_id.strip())
            if turn is None:
                return None
            if turn.status not in TERMINAL_STATUSES:
                turn.status = gestalt.AGENT_EXECUTION_STATUS_CANCELED
                turn.status_message = reason
                turn.completed_at = _utcnow()
                self.append_event(
                    turn_id=turn.turn_id, event_type="turn.canceled", source=turn.provider_name, data={"reason": reason}
                )
            return copy.deepcopy(turn)

    def append_event(self, *, turn_id: str, event_type: str, source: str, data: dict[str, Any]) -> StoredTurnEvent:
        with self._lock:
            events = self._events.setdefault(turn_id, [])
            event = StoredTurnEvent(
                event_id=f"{turn_id}:{len(events) + 1}",
                turn_id=turn_id,
                seq=len(events) + 1,
                event_type=event_type,
                source=source,
                visibility="external",
                data=copy.deepcopy(data),
                created_at=_utcnow(),
            )
            events.append(event)
            return copy.deepcopy(event)

    def list_turn_events(self, *, turn_id: str, after_seq: int = 0, limit: int = 0) -> list[StoredTurnEvent]:
        with self._lock:
            events = [event for event in self._events.get(turn_id.strip(), []) if event.seq > after_seq]
            if limit > 0:
                events = events[:limit]
            return copy.deepcopy(events)


def _utcnow() -> datetime:
    return datetime.now(tz=UTC)


def _assistant_message_event_data(output: gestalt.AgentTurnOutput) -> dict[str, Any]:
    if output.text is not None:
        return {"text": str(output.text or "")}
    structured = output.structured
    if structured is None:
        raise ValueError("completed turn output must include text or structured")
    data: dict[str, Any] = {"text": str(structured.text or "")}
    if structured.value is not None:
        data["value"] = copy.deepcopy(structured.value)
    return data


def session_visibility_for_create(metadata: dict[str, Any], created_by: dict[str, str]) -> str:
    if _is_slack_agent_session_metadata(metadata) and _is_managed_actor(created_by):
        return SESSION_VISIBILITY_COMPANY
    return SESSION_VISIBILITY_PRIVATE


def session_readable_by(session: StoredSession, subject_id: str) -> bool:
    subject_id = subject_id.strip()
    if not subject_id:
        return False
    if str(session.created_by.get("subject_id", "") or "").strip() == subject_id:
        return True
    return session.visibility == SESSION_VISIBILITY_COMPANY


def session_writable_by(session: StoredSession, subject_id: str) -> bool:
    subject_id = subject_id.strip()
    if not subject_id:
        return False
    return str(session.created_by.get("subject_id", "") or "").strip() == subject_id


def _is_slack_agent_session_metadata(metadata: dict[str, Any]) -> bool:
    slack = metadata.get("slack")
    if not isinstance(slack, dict):
        return False
    return bool(
        str(slack.get("team_id") or "").strip()
        and str(slack.get("channel_id") or "").strip()
        and str(slack.get("root_message_ts") or "").strip()
        and str(slack.get("session_ref") or "").strip()
    )


def _is_managed_actor(actor: dict[str, str]) -> bool:
    subject_id = str(actor.get("subject_id", "") or "").strip()
    subject_kind = str(actor.get("subject_kind", "") or "").strip()
    return subject_kind == "service_account" or subject_id.startswith("service_account:")
